polkadot/node/core/rewards-statistics-collector/src/availability_distribution_metrics.rs

```diff
@@ -0,0 +1,141 @@
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::{PerSessionView, View};
+use gum::CandidateHash;
+use polkadot_primitives::{AuthorityDiscoveryId, SessionIndex, ValidatorIndex};
+use std::{
+	collections::{hash_map::Entry, HashMap, HashSet},
+	ops::Add,
+};
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct AvailabilityChunks {
+	pub downloads_per_candidate: HashMap<CandidateHash, HashMap<ValidatorIndex, u64>>,
+	pub uploads_per_candidate: HashMap<CandidateHash, HashMap<ValidatorIndex, u64>>,
+}
+
+impl AvailabilityChunks {
+	pub fn new() -> Self {
+		Self {
+			downloads_per_candidate: Default::default(),
+			uploads_per_candidate: Default::default(),
+		}
+	}
+
+	pub fn note_candidate_chunk_downloaded(
+		&mut self,
+		candidate_hash: CandidateHash,
+		validator_index: ValidatorIndex,
+		count: u64,
+	) {
+		let validator_downloads = self
+			.downloads_per_candidate
+			.entry(candidate_hash)
+			.or_default()
+			.entry(validator_index);
+
+		match validator_downloads {
+			Entry::Occupied(mut validator_downloads) => {
+				*validator_downloads.get_mut() += count;
+			},
+			Entry::Vacant(entry) => {
+				entry.insert(count);
+			},
+		}
+	}
+
+	pub fn note_candidate_chunk_uploaded(
+		&mut self,
+		candidate_hash: CandidateHash,
+		validator_index: ValidatorIndex,
+		count: u64,
+	) {
+		let validator_uploads = self
+			.uploads_per_candidate
+			.entry(candidate_hash)
+			.or_default()
+			.entry(validator_index);
+
+		match validator_uploads {
+			Entry::Occupied(mut validator_uploads) => {
+				*validator_uploads.get_mut() += count;
+			},
+			Entry::Vacant(entry) => {
+				entry.insert(count);
+			},
+		}
+	}
+}
+
+// whenever chunks are acquired throughout availability
+// recovery we collect the metrics about what validator
+// provided and the amount of chunks
+pub fn handle_chunks_downloaded(
+	view: &mut View,
+	session_index: SessionIndex,
+	candidate_hash: CandidateHash,
+	downloads: HashMap<ValidatorIndex, u64>,
+) {
+	let av_chunks = view
+		.availability_chunks
+		.entry(session_index)
+		.or_insert(AvailabilityChunks::new());
+
+	for (validator_index, download_count) in downloads {
+		av_chunks.note_candidate_chunk_downloaded(candidate_hash, validator_index, download_count)
+	}
+}
+
+// handle_chunk_uploaded receive the authority ids of the peer
+// it just uploaded the candidate hash, to collect this statistic
+// it needs to find the validator index that is bounded to any of the
+// authority id, from the oldest to newest session.
+pub fn handle_chunk_uploaded(
+	view: &mut View,
+	candidate_hash: CandidateHash,
+	authority_ids: HashSet<AuthorityDiscoveryId>,
+) {
+	let mut sessions: Vec<(&SessionIndex, &PerSessionView)> = view.per_session.iter().collect();
+	sessions.sort_by(|(a, _), (b, _)| a.partial_cmp(&b).unwrap());
+
+	for (session_idx, session_view) in sessions {
```

Comment:
Going through all sessions seems like it should be unnecessary - we usually know from what session things are coming from.

Reply:
Worse, you might even end up using the wrong session.